website/src/panel-layout.tsx

```diff
@@ -0,0 +1,267 @@
+        <props.panelType word={props.word} features={["englishGloss"]} />
+      )}
+    </>
+  )
+
+  // Contains a component rendering a word's commentary.
+  const commentaryContent = (
+    <props.panelType word={props.word} features={["commentary"]} />
+  )
+
+  return (
+    <>
+      {/* Renders audio recording. */}
+      {/* TODO: CAUSES ERRORS */}
+      <AudioPanel segment={props.word} />
+
+      <CollapsiblePanel
+        title={"Phonetics"}
+        content={phoneticsContent}
+        icon={
+          <MdRecordVoiceOver
+            size={24}
+            className={css.wordPanelButton.colpleft}
+          />
+        }
+      />
+
+      {/* If there are no segments, does not display Word Parts panel */}
+      {props.word.segments.length > 0 && (
+        <CollapsiblePanel
+          title={"Word Parts"}
+          content={wordPartsContent}
+          icon={
+            <IoEllipsisHorizontalCircle
+              size={24}
+              className={css.wordPanelButton.colpleft}
+            />
+          }
+        />
+      )}
+
+      {/* If there is no commentary, does not display Commentary panel */}
+      {props.word.commentary && (
+        <CollapsiblePanel
+          title={"Commentary"}
+          content={commentaryContent}
+          icon={<MdNotes size={24} className={css.wordPanelButton.colpleft} />}
+        />
+      )}
+    </>
+  )
+}
+
+export const VerticalMorphemicSegmentation = (p: {
+  panelType: typeof EditWordPanel | typeof WordPanel
+  segments: FormFieldsFragment["segments"]
+}) => {
+  if (!p.segments) {
+    return null
+  }
+
+  let segmentCount = p.segments.length
+
+  return (
+    <table className={css.tableContainer}>
+      <tbody>
+        {p.panelType === WordPanel
+          ? p.segments.map((segment, index) => (
+              <tr key={index}>
+                <td className={css.tableCells}>
+                  {index > 0 ? "-" : null}
+                  {segment.morpheme}
+                  {index !== segmentCount - 1 ? segment.nextSeparator : null}
+                </td>
+                <td className={css.tableCells}>
+                  {segment.matchingTag
+                    ? segment.matchingTag.title
+                    : segment.gloss}
+                </td>
+              </tr>
+            ))
+          : p.segments.map((segment, index) => (
+              <tr key={index} style={{ display: "flex" }}>
+                <td className={css.editMorphemeCells}>
+                  <p.panelType
+                    features={["segments", index.toString(), "morpheme"]}
+                  />
+                </td>
+                <td className={css.editMatchingCells}>
+                  {segment.matchingTag?.__typename === "TagForm" ? (
+                    <p.panelType
+                      features={[
+                        "segments",
+                        index.toString(),
+                        "matchingTag",
+                        "title",
+                      ]}
```

Comment:
This is a really interesting solution, can you give me a rundown of the thinking behind it? I understand you're looking for a generic component to edit a field at any level of nesting, but I'm curious to hear more. How does this compare to passing a setter like `(word, value) => word.segments[index].matchingTag.title = value` or whatever?

Reply:
I'm not sure if I understood the question 100%, but this wasn't my preferred solution, just one that worked and made the most sense to me. Since Reakit forms display a FormInput of a value with the value's key as a string, I would pass in the value's key as a list of string to identify the desired FormInput. I'm open to other solutions though as this pattern likely won't work now that I'm understanding matching tags correctly!